jobrunner/run.py

```diff
@@ -195,28 +161,161 @@ def handle_running_job(job):
+                # it has failed. If we have an internal error, a full restart
+                # might recover better.
+                raise
+
+
+# we do not control the tranisition from these states, the executor does
+STABLE_STATES = [
+    ExecutorState.PREPARING,
+    ExecutorState.EXECUTING,
+    ExecutorState.FINALIZING,
+]
+
+
+def handle_job_api(job, api):
+    """Handle an active job.
+
+    This contains the main state machine logic for a job. For the most part,
+    state transitions follow the same logic, which is abstracted. Some
+    transitions require special logic, mainly the initial and final states, as
+    well as supporting cancellation.
+    """
+    assert job.state in (State.PENDING, State.RUNNING)
+    definition = job_to_job_definition(job)
+
     if job.cancelled:
-        log.info("Cancellation requested, killing job")
-        api.terminate(job_to_job_definition(job))
+        # cancelled is driven by user request, so is handled explicitly first
+        # regardless of executor state.
+        api.terminate(definition)
+        api.cleanup(definition)
+        mark_job_as_failed(job, "Cancelled by user", StatusCode.CANCELLED_BY_USER)
+        return
 
-    try:
-        is_running = sync_job_status(job, api)
+    old_status = api.get_status(definition)
 
-        if is_running:
-            set_message(job, "Running")
-        else:
-            mark_job_as_completed(job)
-            api.cleanup(job_to_job_definition(job))
-    except JobError as exception:
-        set_message(job, "Failed")
-        mark_job_as_failed(job, exception)
-        api.cleanup(job_to_job_definition(job))
-    except Exception:
-        set_message(job, "Failed")
-        mark_job_as_failed(job, "Internal error when finalising job")
-        # We don't clean up here, to facilitate with debugging this unexpected error
-        raise
+    # handle the simple no change needed states.
+    if old_status.state in STABLE_STATES:
+        # no action needed, simply update job message and timestamp
+        message = old_status.state.value.title()
+        set_message(job, message)
+        return
+
+    # ok, handle the state transitions that are our responsibility
+    if old_status.state == ExecutorState.UNKNOWN:
+        # a new job
+        if job.state == State.RUNNING:
+            log.warning("Got UNKNOWN state for a job we thought was RUNNING")
+
+        # check dependencies
+        awaited_states = get_states_of_awaited_jobs(job)
+        if State.FAILED in awaited_states:
+            mark_job_as_failed(
+                job,
+                "Not starting as dependency failed",
+                code=StatusCode.DEPENDENCY_FAILED,
+            )
+            return
+
+        if any(state != State.SUCCEEDED for state in awaited_states):
+            set_message(
+                job, "Waiting on dependencies", code=StatusCode.WAITING_ON_DEPENDENCIES
+            )
+            return
+
+        next_state = ExecutorState.PREPARING
```

Comment:
Could we call this "expected", instead? I think this better reflects how we actually use it.